ocaml/xapi/message_forwarding.ml

```diff
@@ -1688,10 +1688,15 @@ module Forward = functor(Local: Custom_actions.CUSTOM_ACTIONS) -> struct
              * forward the call to the source. *)
             let snapshot = Db.VM.get_record ~__context ~self:vm in
             fun ~local_fn ~__context ~vm op ->
-              finally
-                (fun () -> allocate_vm_to_host ~__context ~vm ~host ~snapshot ~host_op:`vm_migrate ();
-                           forward_vm_op ~local_fn ~__context ~vm op)
-                (finally_clear_host_operation ~__context ~host ~host_op:`vm_migrate);
+              let res = try
+                allocate_vm_to_host ~__context ~vm ~host ~snapshot ~host_op:`vm_migrate ();
+                forward_vm_op ~local_fn ~__context ~vm op
+              with ex ->
+                debug "migrate_send: allocate_vm_to_host failed. Clearing host operations. Error: %s" (Printexc.to_string ex);
+                finally_clear_host_operation ~__context ~host ~host_op:`vm_migrate ();
+                raise ex in
+              finally_clear_host_operation ~__context ~host ~host_op:`vm_migrate ();
```

Comment:
`finally_clear_host_operation` is called now before the migration takes places, and only called afterwards if there's an error.

I think the previous commit with `finally` would have worked just fine:
If there's an exception in `allocate_vm_to_host` we reset the database state up twice, which is strange, but should get us to an inconsistent state.
If there's an exception within `forward_vm_op` the cleanup happens once.

If there's an exception in `finally_clear_host_operation` the error is logged by `finally` as a warning.

If there are no exceptions it should work just fine.

Unless I'm missing something, using the previous commit would me my preferred solution.

Reply:
i have gone for the 1st commit since the only difference was the `clear_scheduled_to_be_resident_on` call